src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,146 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s._
+import is.hail.asm4s.TypeInfo
+import is.hail.expr.Type
+import is.hail.expr.TStruct
+import is.hail.annotations._
+import org.objectweb.asm.tree._
+import org.objectweb.asm.Opcodes._
+
+import scala.language.existentials
+
+object Compile {
+  case class DetailedTypeInfo[T, UT : TypeInfo](injectNonMissing: Option[(TypeInfo[T], Code[_] => Code[_])]) {
```

Comment:
I don't quite understand this one.

Reply:
I need to know how to inject an unboxed type `UT` into a boxed one `T` when returning from the function.

You mention below that we should say no to boxing everywhere, in that vein, I could force return types to be primitive or reference, but never boxed primitive.